usr/src/cmd/zdb/zdb.c

```diff
@@ -2121,6 +2130,153 @@ dump_label_uberblocks(vdev_label_t *lbl, uint64_t ashift)
 	}
 }
 
+static int
+dump_specific_file(objset_t *os, uint64_t obj)
+{
+	int print_header = 0;
+
+	if (dump_opt['v'] > 5)
+		print_header = 1;
+	dump_object(os, obj, dump_opt['v'], &print_header);
+	return (0);
+}
+
+static int
+dump_specific_directory(objset_t *os, uint64_t obj, char *name)
+{
+	int err;
+	uint64_t child_obj;
+	char *slash = strchr(name, '/');
+	char *namep = name;
+	dmu_buf_t *db;
+	dmu_object_info_t doi;
+
+	if (slash != NULL)
+		*slash = '\0';
+	err = zap_lookup(os, obj, namep, 8, 1, &child_obj);
+	if (slash != NULL)
+		*slash = '/';
+
+	if (err != 0) {
+		fprintf(stderr, "Looking up '%s' failed: %s\n",
+		    name, strerror(err));
+		return (err);
+	}
+
+	child_obj = ZFS_DIRENT_OBJ(child_obj);
+	err = sa_buf_hold(os, child_obj, FTAG, &db);
+	if (err != 0) {
+		fprintf(stderr, "Can't get SA dbuf for obj %lu: %s\n",
+		    obj, strerror(err));
+		return (EINVAL);
+	}
+	dmu_object_info_from_db(db, &doi);
+	sa_buf_rele(db, FTAG);
+
+	if (dump_opt['v'] > 6)
+		fprintf(stderr, "%s: obj %lu name '%s' type %d bonustype %d\n",
+		    __func__, obj, name, doi.doi_type, doi.doi_bonus_type);
+
+	if (doi.doi_bonus_type != DMU_OT_SA &&
+	    doi.doi_bonus_type != DMU_OT_ZNODE) {
+		fprintf(stderr, "Invalid bonus type %d for obj %lu!\n",
+		    doi.doi_bonus_type, obj);
+		return (EINVAL);
+	}
+
+	switch (doi.doi_type) {
+	case DMU_OT_PLAIN_FILE_CONTENTS:
+		ASSERT3P(slash, ==, NULL);
+		return (dump_specific_file(os, child_obj));
+	case DMU_OT_DIRECTORY_CONTENTS:
+		if (slash == NULL) {
+			if (dump_opt['v'] > 6)
+				fprintf(stderr, "Obj %lu name '%s' is a "
+				    "directory at the end of the chain\n",
+				    obj, name);
+			/* Directory is the end of the chain */
+			return (dump_specific_file(os, child_obj));
+		}
+		return (dump_specific_directory(os, child_obj, slash + 1));
+	default:
+		fprintf(stderr, "Object %lu has non-file/directory type %d!\n",
+		    obj, doi.doi_type);
+		return (EINVAL);
+	}
+	/*NOTREACHED*/
+}
+
+static int
+dump_file_path(int argc, char **argv)
+{
+	int err;
+	dsl_dataset_t *ds;
+	char *curp, *last;
+	objset_t *os;
+	uint64_t root_obj;
+
+	/* identifiers: <dataset> <path> */
+	if (argc < 2) {
+		fprintf(stderr, "Must specify dataset and path\n");
+		return (EINVAL);
+	}
+	if (argv[1][0] != '/') {
+		fprintf(stderr, "Must specify absolute path\n");
+		return (EINVAL);
+	}
+	err = dmu_objset_own(argv[0], DMU_OST_ZFS, B_TRUE, FTAG, &os);
+	if (err != 0) {
+		fprintf(stderr, "Can't own dataset '%s': %s\n", argv[0],
+		    strerror(err));
+		return (err);
+	}
+
+	check_sa_attr_table(os);
+	err = zap_lookup(os, MASTER_NODE_OBJ, ZFS_ROOT_OBJ, 8, 1, &root_obj);
+	if (err != 0) {
+		fprintf(stderr, "Can't lookup root znode: %s\n", strerror(err));
+		dmu_objset_disown(os, FTAG);
+		return (EINVAL);
+	}
+	err = dump_specific_directory(os, root_obj, argv[1] + 1);
+	dmu_objset_disown(os, FTAG);
+	sa_attr_table = NULL;
+	return (err);
+}
+
+static enum dmu_object_type
+objtype_from_string(char *objtype_str)
```

Comment:
Ah, I see the flexibility here for adding other object types.  But the names here are pretty hard to type.  What would you think about hard-coding the mapping from "input" names (something simple like "file" or "directory") to dmu_object_type_t?  Or you could add a new field to dmu_object_type_info_t, but that might be overkill.